src/http/http-client.js

```diff
@@ -0,0 +1,218 @@
+                break;
+            }
+        }
+    }
+}
+
+/**
+ * Cancels HTTP requests based on event data from DirectBinaryUploadController's
+ * 'cancel' event data. Depending on event data, the client will either cancel
+ * only those requests whose cancel ID matches the given file name, or cancel
+ * all requests if no fileName is given.
+ *
+ * @param {object} cancelEventData Event data from cancel event.
+ */
+function cancelFromController(cancelEventData) {
+    const { fileName } = cancelEventData;
+
+    if (fileName) {
+        // use fileName as cancel ID and cancel any matches
+        this.cancel(fileName);
+    } else {
+        // no fileName, cancel ALL requests
+        this.cancelAll();
+    }
+}
+
+/**
+ * Supports submitting and managing in-progress HTTP requests. Features of the
+ * client include retrying failed requests, timing request duration, and canceling
+ * in-progress requests.
+ */
+export default class HttpClient extends UploadOptionsBase {
+
+    /**
+     * Constructs a new instance of an HTTP request client using the given
+     * information.
+     * @param {object} options General direct binary upload options.
+     * @param {DirectBinaryUploadOptions} uploadOptions Options specific to
+     *  the current upload.
+     */
+    constructor(options, uploadOptions) {
+        super(options, uploadOptions);
+        this.cancelTokens = {};
+        this.cancelledTokens = {};
+        this.allCancelled = false;
+
+        this.getUploadOptions().getController().on('cancel', cancelFromController.bind(this));
+    }
+
+    /**
+     * Cancels one or more in-progress request(s) by a cancel ID.
+     *
+     * @param {string} cancelId All requests with a matching cancelId
+     *  will be cancelled.
+     */
+    cancel(cancelId) {
+        if (this.cancelTokens[cancelId]) {
+            this.cancelTokens[cancelId].forEach(cancelToken => {
+                cancelToken.cancel(ErrorCodes.USER_CANCELLED);
+            });
+            delete this.cancelTokens[cancelId];
+        }
+        this.cancelledTokens[cancelId] = true;
+    }
+
+    /**
+     * Cancels all in-progress requests, regardless of cancel ID.
+     */
+    cancelAll() {
+        Object.keys(this.cancelTokens).forEach(cancelId => {
+            this.cancel(cancelId);
+        });
+        this.allCancelled = true;
+    }
+
+    /**
+     * Retrieves a value indicating whether a given cancel ID has been
+     * used to cancel an in-progress request. Will return true if the
+     * individual ID was cancelled, or if all requests were cancelled.
+     *
+     * @param {string} cancelId ID to check for cancellation.
+     * @returns {boolean} True if the ID was cancelled, false otherwise.
+     */
+    isCancelled(cancelId) {
+        return this.allCancelled || !!this.cancelledTokens[cancelId];
+    }
+
+    /**
+     * Submits an HTTP request. The client will automatically retry the request if
```

Comment:
Note that for Nui we use https://github.com/adobe/node-fetch-retry.

Reply:
Thanks for the callout. I think switching to the linked module would require a fair amount of work, so I'm going to leave it as-is for now.

As a sidenote, I recently found [exponential-backoff](https://www.npmjs.com/package/exponential-backoff), which I like because it's not tied to a specific HTTP module (i.e. axios, node-fetch, etc). It's not even tied to HTTP for that matter. It's not as widely used as some of the more popular modules, but it worked well when I tried it in a different setting.